### ISSUE TITLE
[no-unnecessary-boolean-literal-compare] Incorrect errors when using optional chaining with indexable types

### ISSUE THREAD

**Issue by gebsh** [typescript-eslint/typescript-eslint#1908]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-boolean-literal-compare": "error"
  }
}
```

```TS
interface Bar {
  [key: string]: { foo: boolean };
}

const bar: Bar = {};

if (bar.baz?.foo === false) { // Error: This expression unnecessarily compares a boolean value (...)
  // Do something only if `baz` exists and `foo` is false.
}
```

**Expected Result**
An error should not be reported.

**Actual Result**
_This expression unnecessarily compares a boolean value to a boolean instead of using it directly_ is reported.

**Additional Info**
Autofix wants to replace this condition with `!bar.baz?.foo` which is incorrect since it's also true when `baz` property does not exist.

The error is not reported when I change `Bar` definition to:
```TS
interface Bar {
  baz?: { foo: boolean };
}
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.28.0` |
| `@typescript-eslint/parser`        | `2.28.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.16.2` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
The rule works off of the types reported by typescript.
In this case - typescript reports that the type of `bar.baz?.foo` is just `boolean`.

But there's an optional chain in there! Where's the `undefined`?
Well it's simple - Typescript does not add `undefined` to a type just because of the optional chain; it will only add `undefined` if the optional chained type is nullish.
Why? Because it's impossible to get `undefined` unless the type is nullish. It would be incorrect types if it just added `undefined` always.

This is why everything starts working when you make the property optional - typescript knows the type is potentially nullish, so the optional chain is potentially `undefined`, and thus the type becomes `boolean | undefined`, and thus our rule sees the type as this.

If you want to keep your map and make it optional, suggest defining your type like this:

```ts
interface Bar {
  [key: string]: { foo: boolean } | undefined;
}
```
